Support HCI_SOCK_USER/RAW

protobits opened the pull request:
## Summary

This is a work in progress to support RAW HCI commands over BTPROTO_HCI socket. As supporting this socket type means that NuttX's host stack is bypassed (packets are sent via socket back to user instead of NuttX's own stack), this PR also adds the ability to disable NuttX's stack entirely. In this case it is expected that a host stack be provided in userspace. 

With this support I have managed to get nimBLE stack working (the demo starts a GATT server which I managed to test from nRF Connect Android app from my smartphone).

Work so far:
* I've extended sendto() implementation for Bluetooth, to distinguish BTPROTO_L2CAP and BTPROTO_HCI
* I've extended the frame metadata struct to include a "protocol" field which allows the radio driver to now which type of
frame it has received
* I've extended btnet_req_data to support each of these cases so that the frame is forwarded to either L2CAP part of stack or directly to bluetooth driver as HCI buffer
* I've added a callback to bt netdev driver which is used in bt_hcicore to forward HCI packets while they are being dequeued back to the network driver
* I've extended bluetooth_connection_s struct and handling to include the notion of protocol, as it is required to consider a non-connected but bound socket when looking where to send a packet received from the controller

I have tested this with NuttX's host layer disabled but I would need to retest with NuttX's host layer enabled to validate that is still working. I would also need to condition the support of HCI RAW channel to when the host layer is disabled. My thinking is that  NuttX would either support: host-layer disabled + HCI RAW channel OR host-layer enabled + BTPROTO_L2CAP + BTPROTO_HCI w/MGMT channel. I think other combinations would not make sense or would be too difficult to support, but I'm not sure.

There's a few hacks which I left for proper solutions which required more work:
* The sockaddr_hci indicates which BT device should receive the packet. I understand this is an index over all BT devices (not ALL network devices) so it requires this search which is not easy. I simply plugged a search for the n-th driver, which works for my test case on sim.
* Error handling in various places

In https://github.com/apache/mynewt-nimble/compare/master...v01d:nuttx you can see the (also WIP) changes to nimBLE to add support for NuttX. It is mostly based on Linux port. This port originally used sendmsg() which in NuttX is not fully supported so I changed it to sendto() which is mostly the same. It uses `read()` for receving, which is strange, but works OK. There are a few fixes I needed to introduce to nimBLE which I hope I can get ported upstream to have proper support.

To be done:
1. Polish this
2. Validate approach to the case of supporting BTPROTO_L2CAP + MGMT
3. More testing

I also think that we need to decide when/how we're going to migrate to a pure socket based interface and drop support for current ioctl() based interface. In that case I think that supporting bluetooth and supporting AF_BLUETOOTH sockets would be just one option instead of two separate ones. This will also break support for btsak until something else which uses the socket interface is written.

## Impact

Adds new support, for now it shouldn't break anything, but it requires testing

## Testing

I'm testing this on sim with hci socket support and nimBLE. nimBLE is being built as an application (the stack + demo "nimble" app). Eventually it will be split into a library and optional demo apps. The test is as follows:

1. boot sim nuttx (with proper permission for HCI access)
2. ifup bnep0
3. nimble

While starting, nimble app start interacting with controller by sending an HCI commands and registers a GATT server. I've connected with my phone and works with no issues.

YuuichiNakamura committed:
Add note_syscall args support

btashton committed:
Bluetooth: Start implementing BTPROTO_HCI socket support

Signed-off-by: Brennan Ashton <bashton@brennanashton.com>

YuuichiNakamura committed:
Add SCHED_INSTRUMENTATION_HIRES support

Commit:
bluetooth: Add support for HCI RAW channel; make host layer optional

v01d commented:
Ignore the changes to Config.mk, it is unrelated

btashton commented on wireless/bluetooth/bt_netdev.c:
```diff
@@ -945,7 +993,79 @@ static int btnet_req_data(FAR struct radio_driver_s *netdev,
+
+      switch (iob->io_data[iob->io_offset])
+        {
+          case BT_H4_CMD:
```
Why are we doing H4 packets here and not:

````
/* HCI data types */
#define HCI_COMMAND_PKT		0x01
#define HCI_ACLDATA_PKT		0x02
#define HCI_SCODATA_PKT		0x03
#define HCI_EVENT_PKT		0x04
#define HCI_VENDOR_PKT		0xff
```

protobits replied:
I think we're talking about the same thing. I see you defined these in the simulated HCI socket driver. I added this definition in netpacket/bluetooth.h. I can put these there and use those names instead.

protobits replied:
I replaced these with your macros which I understand align more to Linux naming

btashton commented on wireless/bluetooth/bt_netdev.c:
```diff
@@ -945,7 +993,79 @@ static int btnet_req_data(FAR struct radio_driver_s *netdev,
+          return -ENOMEM;
+        }
+
+      g_btdev.btdev->send(g_btdev.btdev, buf);
```
This should use `bt_send`

protobits replied:
This was initially like that. bt_send is actually an internal function so I exposed it as public. But I realized this is part of hcicore.c, and this socket would be skipping NuttX's host layer, so I think it is preferable to call this directly. Also, hcicore is the Intel owned code, just to be aware of.

btashton replied:
But we need a way to intercept all of the outgoing HCI transfers before they go to the driver to be captured by monitoring channels.  Also it is wrong to be calling `g_btdev` everywhere so I am trying to cut down on that to make supporting multiple devs easier later.

protobits replied:
Maybe this can be implemented in a new function in bt_netdev.c then. Not sure how to handle the multiple-device case yet.

btashton replied:
My plan for multiple devices was to add a field to the buf structure that contains the btdev pointer.

btashton replied:
This seems wrong to me why are we not putting the data in the tx_queue and having `hci_tx_kthread` do the sending.

This would just be implementing the function you need in `bt_hcicore.c` that looks like `static void hci_cmd_status(FAR struct bt_buf_s *buf)` whith the logic that you need

protobits replied:
That was one of my questions earlier in the other PR: does it make sense to have two queues? I understand the HCI queue was probably there because of how it was integrated to userspace via direct calls. But once the networking layer is the only access point to HCI from userspace, do we really need a queue besides the networking one?

protobits replied:
> This would just be implementing the function you need in `bt_hcicore.c` that looks like `static void hci_cmd_status(FAR struct bt_buf_s *buf)` whith the logic that you need

Sorry, don't understand what you mean

btashton replied:
You still need the kernel to be able to manage the interface (I get you are bypassing it right now). If you send a scan for n sec control request via the management interface the kernel will need to handle things like making sure the stop scan is scheduled correctly.

If we keep the same queue logic it will be a lot easier to remove a layer later if we need to.

protobits replied:
Maybe we need to extract the queue handling from hcicore.c into a different file then. Right now it is a mix of part of the host stack implementation and an API to communicate with the controller. For example, there's both a bt_initialize and hci_initialize there. If you agree I can do that. Sadly this means that even when using nimBLE we would still have Intel owned code built, but I guess we can worry about that later.

btashton replied:
> > This would just be implementing the function you need in `bt_hcicore.c` that looks like `static void hci_cmd_status(FAR struct bt_buf_s *buf)` whith the logic that you need
> 
> Sorry, don't understand what you mean

That was the wrong function sorry.  We should not be calling the driver here we should be calling a higher level API. 
Can we just call `bt_hci_cmd_send`

btashton replied:
> Maybe we need to extract the queue handling from hcicore.c into a different file then. Right now it is a mix of part of the host stack implementation and an API to communicate with the controller. For example, there's both a bt_initialize and hci_initialize there. If you agree I can do that. Sadly this means that even when using nimBLE we would still have Intel owned code built, but I guess we can worry about that later.

I think we can worry about that later.  I think once the GATT logic is remove there will actually be very little left of the Intel code and what is left is interfaces that are clearly inspired from things like bluez.  Carrying a few files with the Intel BSD licence is not a deal breaker, we are going to end up having that anyway in other parts of the OS we just want to minimize it.

My hope would be that we can play with using the `nimBLE` stack where we just leverage the GATT layer on top of the L2CAP sockets, but I understand wanting to first just run it fully in userspace.

protobits replied:
> My hope would be that we can play with using the `nimBLE` stack where we just leverage the GATT layer on top of the L2CAP sockets, but I understand wanting to first just run it fully in userspace.

That will probably require quite some work on nimBLE. At that point we should try to collaborate with mynewt devs to see how they feel about the idea and maybe try to get it integrated.

protobits replied:
> That was the wrong function sorry. We should not be calling the driver here we should be calling a higher level API.
> Can we just call `bt_hci_cmd_send`

`bt_hci_cmd_send` just adds one level of indirection before sending to the queue and the problem is that it requires introspecting the packet since it expects the opcode already. I think it is better to replace bt_send with bt_queue_send, which is what bt_hci_cmd_send uses. I will see if I can separate bt_* functions from hci_* functions and have it separately built so that it is easier to deactivate NuttX's host layer.

protobits replied:
Mmm, bt_queue_send receives the queue as parameter. I think this requires a bit of API rework to not expose the queue.
I'll take note of these changes but I really want to get RX working now that TX is working. I want to test nimBLE itself aswell.

btashton commented on net/bluetooth/bluetooth_sockif.c:
```diff
@@ -904,9 +906,9 @@ static ssize_t bluetooth_sendto(FAR struct socket *psock,
+  /* Only SOCK_RAW is supported */
 
-  if (psock->s_type == SOCK_RAW && psock->s_proto == BTPROTO_L2CAP)
+  if (psock->s_type == SOCK_RAW)
```
We should break this out into the different sockets here like I did for other ones instead of further changing `psock_bluetooth_sendto`  i.e.
`psock_bluetooth_sendto` -> `psock_bluetooth_l2cap_sendto`  `psock_bluetooth_hci_sendto`

There are a lot of difference in these functions especially when you start adding the channels.

protobits replied:
I chose to use just `psock_bluetooth_sendto` for both since right now there's a lot of common code, but if it gets bigger with the other channels we can split it.

protobits commented on net/socket/sendmsg.c:
```diff
@@ -93,19 +93,14 @@ ssize_t psock_sendmsg(FAR struct socket *psock, FAR struct msghdr *msg,
       return -EINVAL;
     }
 
-  if (msg->msg_iovlen != 1)
```
This shouldn't actually be here. I initially wanted to use sendmsg() and extend it but I decided to just use sendto() instead.

v01d commented:
@btashton I worked a bit more on the RX part and though the best way would be to add a callback to bt_netdev.c which bt_hcicore.c can invoke from within the `priority_rx_work` function, to pass an HCI packet to the driver. From there, it follows a similar path to that of received L2CAP packets, calling `bluetooth_input` to send the packet to the appropriate socket.

After some other changes I now need one final thing to complete this: I need to prepend the H4 header (I believe that is the name, the one byte header indicating CMD, EVT, etc.) to the HCI data received as bt_buf_s, to obtain a iob_s containing this header + packet. Do you know how this is supposed to be done? This is what is done for L2CAP in `btnet_l2cap_receive` case before calling `bluetooth_input`:

```
  frame      = buf->frame;
  buf->frame = NULL;

  /* Ignore the frame if the network is not up */

  priv = (FAR struct btnet_driver_s *)context;
  if (!priv->bd_bifup)
    {
      wlwarn("WARNING: Dropped... Network is down\n");
      goto drop;
    }

  /* Make sure that the size/offset data matches the buffer structure data.
   * REVISIT:  Wouldn't it be better to just have one copy rather than having
   * to synchronize?
   */

  frame->io_len    = buf->len;
  frame->io_pktlen = buf->len;
  frame->io_offset = (unsigned int)
    ((uintptr_t)buf->data - (uintptr_t)frame->io_data);
```

v01d commented:
Ok, I got it. Wasn't too hard after all. 
Now nimBLE appears to be receiving the packet. I need to continue debugging to see if there aren't other things missing.

v01d commented:
Hey! It's working!
I just paired my smartphone =D

I had to fix a race condition in nimble but other than that it appears to be working. I will push what I have to both repos and see about getting it to a mergeable state.

v01d commented:
![Peek 2020-09-21 18-47](https://user-images.githubusercontent.com/161706/93825144-0d062d00-fc3b-11ea-9468-da074d66e5a3.gif)

v01d commented:
@btashton I've updated the PR description to reflect the recent changes (which I added as separate commits to help reviewing)
I'm interested in your perspective for the case of supporting BTPROTO_L2CAP+MGMT channel. I tried to be aware of this case but I may have missed it somewhere.

btashton commented:
@v01d sorry for the little delay, I'll take a look at it tonight.  My biggest concern is that with this we are moving the whole stack into userspace which means you cannnot use the l2cap, rfcomm, etc... with this.   This is why I thought we were talking about using nimble to provide the layer on top of l2cap and then make the calls the the management socket. I get that it is a lot of changes, but that is the only way I could see to get this in in an official way.

v01d commented:
> @v01d sorry for the little delay, I'll take a look at it tonight. My biggest concern is that with this we are moving the whole stack into userspace which means you cannnot use the l2cap, rfcomm, etc... with this. This is why I thought we were talking about using nimble to provide the layer on top of l2cap and then make the calls the the management socket. I get that it is a lot of changes, but that is the only way I could see to get this in in an official way.

NuttX's stack remains in place and can be activated or not via Kconfig (with the option I added). If you disable it, you can provide your own stack, which in this case would be nimBLE as an application/library interacting via RAW sockets. If you leave it enabled, it should work as before and could be modified to use L2CAP socket + mgmt socket as you intended. So, these approaches are not incompatible.

This could be a first step. A second step could be to try to talk with nimBLE devs if they would be open to add support for allowing their L2CAP stack be disabled and use it via L2CAP sockets instead. However, I would only do that if our intention were to replace NuttX stack with nimBLE, but I think that requires further community discussion. A good first step would be to support both stacks and then decide if we want to switch or continue with current stack as the official one.

My aim is simply to provide nimBLE as an external option. I don't think this needs to conflict with the work to support L2CAP + mgmt. We should simply try to iterate this to support both configurations.

Finally, regarding USER vs RAW for nimBLE. My expectation is as follows: when using nimBLE, NuttX's stack is disabled. However, each app would access the stack independently so using RAW would allow this. Using USER means that we need something like a central access point to the stack, via a daemon for example. I'd avoid that approach if possible.

btashton commented:
> > @v01d sorry for the little delay, I'll take a look at it tonight. My biggest concern is that with this we are moving the whole stack into userspace which means you cannnot use the l2cap, rfcomm, etc... with this. This is why I thought we were talking about using nimble to provide the layer on top of l2cap and then make the calls the the management socket. I get that it is a lot of changes, but that is the only way I could see to get this in in an official way.
> 
> NuttX's stack remains in place and can be activated or not via Kconfig (with the option I added). If you disable it, you can provide your own stack, which in this case would be nimBLE as an application/library interacting via RAW sockets. If you leave it enabled, it should work as before and could be modified to use L2CAP socket + mgmt socket as you intended. So, these approaches are not incompatible.
> 
> This could be a first step. A second step could be to try to talk with nimBLE devs if they would be open to add support for allowing their L2CAP stack be disabled and use it via L2CAP sockets instead. However, I would only do that if our intention were to replace NuttX stack with nimBLE, but I think that requires further community discussion. A good first step would be to support both stacks and then decide if we want to switch or continue with current stack as the official one.
> 
> My aim is simply to provide nimBLE as an external option. I don't think this needs to conflict with the work to support L2CAP + mgmt. We should simply try to iterate this to support both configurations.
> 

I think we are mostly in agreement on this part.

> Finally, regarding USER vs RAW for nimBLE. My expectation is as follows: when using nimBLE, NuttX's stack is disabled. However, each app would access the stack independently so using RAW would allow this. Using USER means that we need something like a central access point to the stack, via a daemon for example. I'd avoid that approach if possible.

How do you see that working. What socket inteface/how would these apps talk? I don't see how RAW vs USER has anything to do with this.  You cannot have two applications sending HCI commands or you will have problems keeping the state in sync. This is why nimBLE uses USER so it has exclusive rights to manage HCI on the wire.

v01d commented:
> How do you see that working. What socket inteface/how would these apps talk? I don't see how RAW vs USER has anything to do with this. You cannot have two applications sending HCI commands or you will have problems keeping the state in sync. This is why nimBLE uses USER so it has exclusive rights to manage HCI on the wire.

I was thinking mostly about introspection tools, mostly. But, yet, two apps sending HCI commands that alter state would make no sense. Maybe it is better to guarantee exclusive access via USER for this case. In that scenario, using nimBLE would imply to a single app should have access to the stack. I don't think that sounds bad for a first approach.

Let me know later when you look at the changes closer. I think the goal would be to test both configuration: nuttx host stack + L2CAP + mgmt versus nimBLE stack + USER channel. With that I think we could make a first merge to NuttX, which shouldn't break anything. After that and some more testing we could open the discussion on which stack looks better to embrace.

Donny9 committed:
sensor: add wtgahrs2 init and enable config

Change-Id: I0521098d6157c59d4e0d463e43a2d202797577b6
Signed-off-by: dongjiuzhu <dongjiuzhu1@xiaomi.com>

v01d commented:
hi @btashton , did you have time to take a closer look to this? I would really like to get the RAW HCI socket merged in NuttX so that I can then make a PR to nimBLE repo adding support for NuttX. I have tested this combination in sim and partially on nRF52 (with my WIP link-layer) and it works really well. 

Do you think that, after a bit of more iteration to this PR, we could merge this to mainline? L2CAP_PROTO + mgmt can be added in a second step. I can make the switch to the USER channel so that only exclusive access is supported when the whole stack is in userspace (as with nimBLE). Did you had any other changes in mind?

btashton commented:
I will look at it this weekend. Sorry it was my first week on a new job so I have not had much time and I want to give a detailed review that is deserves.

I just want to make sure we are using a solid API boundary and not hacking this in.

Longer term I think it's not a good idea to keep this only in user space since it breaks all the other expected functionality via sockets, so I want to be sure we are not crippling that. That still means we can fully support this as you have done and does not block this.

We should also consider if we want to put this in the 10.0.0 release or wait another cycle (merge right after the release). I'm inclined to do that just so we have more time to tweak the API and do testing.

v01d commented:
> I will look at it this weekend. Sorry it was my first week on a new job so I have not had much time and I want to give a detailed review that is deserves.

Sure! Just wanted to touch base.

> I just want to make sure we are using a solid API boundary and not hacking this in.
> 
> Longer term I think it's not a good idea to keep this only in user space since it breaks all the other expected functionality via sockets, so I want to be sure we are not crippling that. That still means we can fully support this as you have done and does not block this.

I don't really see an incompatibility with both approaches, I added the option to disable NuttX host layer for this reason. RAW access would only be provided in this case. In the opposite case, L2CAP + mgmt would be supported. If we can support both options without conflict I think it is a good option to have.

But I agree that there should be a discussion soon about what we want to do with NuttX host stack: should we continue to maintain the current one or should we use nimBLE (or even something else) as our official one?

> We should also consider if we want to put this in the 10.0.0 release or wait another cycle (merge right after the release branch). I'm inclined to do that just so we have more time to tweak the API and do testing.

Yeah, there's no rush, it is better to have 10.0 use current approach.

Donny9 committed:
epoll_create: mofidy return value for epoll_create1

Change-Id: I422d7640c37105e1120db18b29d56547f447846f
Signed-off-by: dongjiuzhu <dongjiuzhu1@xiaomi.com>

btashton commented:
@v01d I apologise for the delay in reviewing. I have set some time aside tonight.

v01d commented:
No problem! Thank you!

v01d commented:
Hi @btashton, I'm thinking... would you rather I open this PR to nuttx mainline? This way we can get other reviewers in and your are not overburdened with this (I understand you are busy with next release and new job). If so, the PR would include both your initial changes and mine. There's some iteration I still would have to do to switch to USER channel instead of RAW though (so it would be a draft PR in the meantime), but I think it is a good first self-contained change (before addressing the L2CAP+MGMT you intended to do) .

liuguo09 committed:
arch: Add custom arch chip build support

Just like custom board build support, add custom arch chip build
support.

Change-Id: I71c87e6b2195501a1b1d728b71d7cbe344951057
Signed-off-by: liuhaitao <liuhaitao@xiaomi.com>

xiaoxiang781216 committed:
libc: Don't fclose and fopen file in freopen

to avoid FILE pointer change after this commit:
commit b0797263ca2817cf536be8b843ac5b5aac5d4d9c
Author: Xiang Xiao <xiaoxiang@xiaomi.com>
Date:   Thu Aug 13 18:17:29 2020 +0800

    libc/stdio: Allocate file_struct dynamically

    1.Reduce the default size of task_group_s(~512B each task)
    2.Scale better between simple and complex application

Signed-off-by: Xiang Xiao <xiaoxiang@xiaomi.com>

xiaoxiang781216 committed:
sched: Call up_[use|create]_stack after nxtask_setup_scheduler

to ensure the basic info(e.g. pid) setup correctly before call arch API

Signed-off-by: Xiang Xiao <xiaoxiang@xiaomi.com>
Change-Id: I851cb0fdf22f45844938dafc5981b3f576100dba

YuuichiNakamura committed:
Update trace command document

diegoherranz committed:
Documentation/quickstart/debugging.rst: rewrite JTAG/OpenOCD section

- Add mentions to SWD
- Sony's OpenOCD fork no longer needed since Nuttx support was merged upstream
on http://openocd.zylin.com/#/c/4103/. Recommended to build from source or use
one of the unofficial, more recent builds.
- Amazon link for Olimex adapter was no longer available so removed.
- Reordered and some parts rewritten

xiaoxiang781216 committed:
build: Change .external_dummy to dummy

to better align with the arch, boards and drivers naming convention

Signed-off-by: Xiang Xiao <xiaoxiang@xiaomi.com>

masayuki2009 committed:
arch: imx6: Fix style warnings in chip.h

Signed-off-by: Masayuki Ishikawa <Masayuki.Ishikawa@jp.sony.com>

xiaoxiang781216 committed:
Change all 'Nuttx' to 'NuttX'

Unify the naming convention

Signed-off-by: Xiang Xiao <xiaoxiang@xiaomi.com>

xiaoxiang781216 committed:
Fix nxstyle warning

Signed-off-by: Xiang Xiao <xiaoxiang@xiaomi.com>

diegoherranz committed:
Documentation/quickstart/debugging.rst: fix NuttX capitalisation

AlexanderVasiljev committed:
serial: break from read after closing

rzr committed:
freedom-k64f: Add demo configuration

Minimal configuration.

Each feature will added one after one
to track configuration flags.

Change-Id: Iab8aff1d489a518b6050ecba0f0b950a8b27745b
Bug: https://github.com/apache/incubator-nuttx/issues/1988
Forwarded: https://github.com/apache/incubator-nuttx/pull/2037
Signed-off-by: Philippe Coval <rzr@users.sf.net>

masayuki2009 committed:
arch: arm: Fix interrupt stack handlings for SMP

Summary:
- Modify arm_intstack_base() to return "top" of the IRQ stack for the current CPU
- This change fixes IRQ stack dump information for ARM SMP
- Add arm_intstack_alloc() to return "bottom" of the IRQ stack for the current CPU
- Also, these functions are now implemented in xxx_irq.c (imx/cxd56/lc823450)
- up_color_intstack() and up_check_intstack() now call arm_intstack_alloc()
- These semantics are now consistent with non-SMP case
- up_color_intstack() now initializes whole IRQ stack region for SMP
- Adjust IRQ stack top address for each CPU (e.g. -8)
- Fix setintstack to handle in case of NCPUS=1 (cxd56, lc823450)
- Adjust INTSTACK_SIZE to 8 bytes alignment (cxd56, lc823450)
- Refactor setintstack for lc823450
- Remove old IRQ stack coloring code from up_irqinitialize() (lc823450)
- Introduce g_cpu_intstack_top for lc823450
- Refactor header files

Impact:
- Affects imx6/cxd56xx/lc823450 SMP with interrupt stack enabled

Testing:
- Tested with sabre-6quad:smp (with QEMU, NCPUS=1 and 4)
- Tested with spresense:wifi_smp (NCPUS=1 and 2)
- Tested with lc823450-xgevk:rndis (NCPUS=1 and 2)

Signed-off-by: Masayuki Ishikawa <Masayuki.Ishikawa@jp.sony.com>

Donny9 committed:
driver/sensor: add unified management for sensor

Signed-off-by: dongjiuzhu <dongjiuzhu1@xiaomi.com>

Donny9 committed:
sensor/driver: wtgahrs2 by serial interface follow sensor.c/sensor.h

Wtgahrs2 integrates multiple sensor: accel, gyro, mag, baro and gps.

Signed-off-by: dongjiuzhu <dongjiuzhu1@xiaomi.com>

rzr committed:
freedom-k64f: Regenerate demo/defconfig

With default settings,
it has been verified using:

    ./tools/refresh.sh --silent freedom-k64f/demo

Change-Id: I3cb96b82a392ac169d7887f5afda37a8b5daf992
Bug: https://github.com/apache/incubator-nuttx/issues/1988
Forwarded: https://github.com/apache/incubator-nuttx/pull/2037
Signed-off-by: Philippe Coval <rzr@users.sf.net>

hartmannathan committed:
tiva/cc13x0: Fix nxstyle errors

arch/arm/src/tiva/hardware/cc13x0/cc13x0_adi4_aux.h,
arch/arm/src/tiva/hardware/cc13x0/cc13x0_aux_smph.h,
arch/arm/src/tiva/hardware/cc13x0/cc13x0_ccfg.h,
arch/arm/src/tiva/hardware/cc13x0/cc13x0_ddi.h,
arch/arm/src/tiva/hardware/cc13x0/cc13x0_memorymap.h,
arch/arm/src/tiva/hardware/cc13x0/cc13x0_smph.h:

    * Fix nxstyle errors.

juniskane committed:
arch: serial: fix all TCGETS retrieving zero baud rate

cfsetispeed() now stores baud rate to c_cflag member of
struct termios, so it must not be overridden later on.

Signed-off-by: Juha Niskanen <juha.niskanen@haltian.com>

yamt committed:
arch/xtensa/src/esp32/esp32_wlan.c: Fix an unused var warning

chip/esp32_wlan.c: In function 'esp_ioctl':
chip/esp32_wlan.c:1262:30: warning: unused variable 'req' [-Wunused-variable]
   struct mii_ioctl_notify_s *req;
                              ^~~

masayuki2009 committed:
boards: lc823450-xgevk: Update rndis/defconfig

Summary:
- Add CONFIG_ARCH_STACKDUMP=y
- Add CONFIG_STACK_COLORATION=y
- Remove CONFIG_NSH_DISABLE_MB=y
- Remove CONFIG_NSH_DISABLE_MH=y

Impact:
- Affects lc823450-xgevk:rndis only

Testing:
- Tested with smp, ostest.

Signed-off-by: Masayuki Ishikawa <Masayuki.Ishikawa@jp.sony.com>

masayuki2009 committed:
boards: sabre-6quad: Update smp/defconfig

Summary:
- Remove CONFIG_HOST_WINDOWS=y
- Add CONFIG_READLINE_CMD_HISTORY=y
- Add CONFIG_STACK_COLORATION=y

Impact:
- Affects sabre-6quad:smp only

Testing:
- Tested with smp, ostest with QEMU

Signed-off-by: Masayuki Ishikawa <Masayuki.Ishikawa@jp.sony.com>

btashton committed:
serial: nxstyle fixes

Signed-off-by: Brennan Ashton <bashton@brennanashton.com>

btashton committed:
Docs: Bump version of pygments to get new GAS lexer

Signed-off-by: Brennan Ashton <bashton@brennanashton.com>

btashton committed:
Docs: Fix broken links and references in docs

Also add to the CI a problem matcher to better report issues

Signed-off-by: Brennan Ashton <bashton@brennanashton.com>

yamt committed:
esp32: Fix a few #endif comments

rzr committed:
freedom-k64f: Add I2C to demo image

Both flag were selected manualy from menuconfig

Usage:

    nsh> ls /dev/i2c0

    nsh> i2c bus
    BUS   EXISTS?
    Bus 0: YES

    nsh> i2c dev 0 0x7F
         0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
    00: 00 -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
    10: -- -- -- -- -- -- -- -- -- -- -- -- -- 1d -- --

Change-Id: I86c68753609c521af9710190397ef492ae562e38
Bug: https://github.com/apache/incubator-nuttx/pull/1999
Forwarded: # https://github.com/apache/incubator-nuttx/pull/2047
Signed-off-by: Philippe Coval <rzr@users.sf.net>

saramonteiro committed:
xtensa/tools: Updated makefile to download the IDF stable version

rzr committed:
kinetis: Make kinetis_i2c_sem_wait_noncancelable optional

CONFIG_I2C_RESET is coming from imxrt_lpi2c.c

Change-Id: I8f55eaf793386fe5ac1d4adaf310d6e0f08dcd92
Bug: https://github.com/apache/incubator-nuttx/pull/1999
Forwarded: https://github.com/apache/incubator-nuttx/pulls/rzr
Signed-off-by: Philippe Coval <rzr@users.sf.net>

Donny9 committed:
serial/termios: support custom baud rate setting

N/A

Change-Id: I0cc00f6d0fda852533bd4a38ea6fc39e0d0059da
Signed-off-by: dongjiuzhu <dongjiuzhu1@xiaomi.com>

liuguo09 committed:
configure.sh: update custom board config build

If Make.defs not found under boardconfig dir, then try
scripts/Make.defs.

Signed-off-by: liuhaitao <liuhaitao@xiaomi.com>

diegoherranz committed:
Documentation/quickstart/debugging: rewrite SWD sentence

Original content rather than copy-paste from Wikipedia.
Discussed on https://github.com/apache/incubator-nuttx/pull/2031

rzr committed:
freedom-k64f/demo: Enable CONFIG_SENSORS

This change is mostly for testing codebase

Change-Id: I50a024dd80661932bfb06aed5be2b4addbbdee8a
Bug: https://github.com/apache/incubator-nuttx/pull/1999
Forwarded: https://github.com/apache/incubator-nuttx/pulls/rzr
Signed-off-by: Philippe Coval <rzr@users.sf.net>

hartmannathan committed:
tiva/cc13x2_cc26x2: Fix nxstyle errors

arch/arm/src/tiva/hardware/cc13x2_cc26x2/cc13x2_cc26x2_adi4_aux.h,
arch/arm/src/tiva/hardware/cc13x2_cc26x2/cc13x2_cc26x2_aux_smph.h,
arch/arm/src/tiva/hardware/cc13x2_cc26x2/cc13x2_cc26x2_ddi.h,
arch/arm/src/tiva/hardware/cc13x2_cc26x2/cc13x2_cc26x2_memorymap.h,
arch/arm/src/tiva/hardware/cc13x2_cc26x2/cc13x2_cc26x2_smph.h:

    * Fix nxstyle errors.

rzr committed:
freedom-k64f: Enable SENSORS_FXOS8700CQ in demo config

Testing this flag in the CI pipeline

Change-Id: I46bf0d6f13d1d45e196e69733776f002f080e2ff
Bug: https://github.com/apache/incubator-nuttx/pull/1999
Forwarded: https://github.com/apache/incubator-nuttx/pulls/rzr
Signed-off-by: Philippe Coval <rzr@users.sf.net>

yamt committed:
math: Make this friendly with libcxx

- Turn some macros into functions

- Implement some type-agnostic functions.
  (Just use __builtin_xxx)

- Add some missing function prototypes
  (Just prototypes, not actually implemented in this commit)

masayuki2009 committed:
Revert "sched: Call up_[use|create]_stack after nxtask_setup_scheduler"

This reverts commit 6eba1ad90a3f49fc085e280d5e517a4656689732.

rzr committed:
freedom-k64f/demo: Enable EXAMPLES_FXOS8700CQ

Depends on other flags that will be enabled automatically by kconfig:

  CONFIG_SENSORS=y
  CONFIG_SENSORS_FXOS8700CQ=y

Change-Id: Ic9bfbd02cca9f2fd4b1f1fb538100865a1c497ff
Bug: https://github.com/apache/incubator-nuttx/pull/1999
Forwarded: https://github.com/apache/incubator-nuttx/pulls/rzr
Signed-off-by: Philippe Coval <rzr@users.sf.net>

rzr committed:
freedom-k64f/demo: Support Ethernet

It was tested using:

  nsh> ifconfig
  eth0    Link encap:Ethernet HWaddr 52:13:FF:FF:FF:FF at DOWN
          inet addr:10.0.0.2 DRaddr:10.0.0.1 Mask:255.255.255.0

All flags are needed to prevent build failures like:

    chip/kinetis_enet.c:181:4:\
    error: #error "Unrecognized or missing PHY selection"

    netlib_setifstatus.c: In function 'netlib_ifup':
    netlib_setifstatus.c:83:27: \
    error: 'NETLIB_SOCK_TYPE' undeclared (first use in this function);

Change-Id: I9496528ae09f6244d73ab77dbcd307ced95f46fd
Bug: https://github.com/apache/incubator-nuttx/pull/1988
Forwarded: https://github.com/apache/incubator-nuttx/pull/2066
Signed-off-by: Philippe Coval <rzr@users.sf.net>

rzr committed:
freedom-k64f/demo: Support DHCP renew tool

Usage:

    nsh> ifconfig

    eth0    Link encap:Ethernet HWaddr 52:13:FF:FF:FF:FF at UP
            inet addr:10.0.0.2 DRaddr:10.0.0.1 Mask:255.255.255.0

    nsh> renew eth0 && ifconfig

    eth0    Link encap:Ethernet HWaddr 52:13:FF:FF:FF:FF at UP
            inet addr:192.168.1.42 DRaddr:192.168.1.254 Mask:255.255.255.0

CONFIG_NETDB_DNSCLIENT is needed for

    apps/netutils/netinit/netinit.c:659: \
    undefined reference to `netlib_set_ipv4dnsaddr'

Change-Id: I54b8495637baf0f3f28abf67d4d4ce861ef49ced
Bug: https://github.com/apache/incubator-nuttx/issues/1988
Forwarded: https://github.com/apache/incubator-nuttx/pulls/rzr
Signed-off-by: Philippe Coval <rzr@users.sf.net>

hartmannathan committed:
tiva/cc13x2_cc26x2: Fix nxstyle errors

arch/arm/src/tiva/hardware/cc13x2_cc26x2/cc13x2_cc26x2_ccfg.h,
arch/arm/src/tiva/hardware/cc13x2_cc26x2/cc13x2_cc26x2_vims.h:

    * Fix nxstyle errors.

juniskane committed:
stm32: common: fix SSD1306 initialization when LED is in SPI bus

Signed-off-by: Juha Niskanen <juha.niskanen@haltian.com>

juniskane committed:
board: LCD: fix wrong number of parameters to lcderr() macro

Signed-off-by: Juha Niskanen <juha.niskanen@haltian.com>

Bhindhiya committed:
RX65N ipv6 defconfig modified

sgysh committed:
sched/signal/sig_procmask.c: Fix a typo

sgysh committed:
sched/signal/sig_procmask.c: Remove an unnecessary variable

sgysh committed:
Remove tabs and spaces at the end of lines

sgysh committed:
Fix nxstyle warnings

Commit:
bt_netdev: remove invalid assert breaking build when debugging is enabled

Commit:
Add $(OBJS) and $(BIN) to CLEAN macro to clean objects also in subdirectories

sgysh committed:
libs/libc/string/lib_strsignal.c: Fix a comment

Commit:
bt_hcicore.c: Fix wrong order of bt_send() and setting last command sent

Commit:
documentation: change non-working URL
